src/test/func/sandbox/sandbox.cc

```diff
@@ -0,0 +1,234 @@
+       */
+      MemoryProviderStateMixin<NoOpPal>* real_state;
+
+      /**
+       * Pop an element from the large stack for the specified size class,
+       * proxies to the real implementation.
+       *
+       * This method must be implemented for `LargeAlloc` to work.
+       */
+      void* pop_large_stack(size_t large_class)
+      {
+        return real_state->pop_large_stack(large_class);
+      };
+
+      /**
+       * Push an element to the large stack for the specified size class,
+       * proxies to the real implementation.
+       *
+       * This method must be implemented for `LargeAlloc` to work.
+       */
+      void push_large_stack(Largeslab* slab, size_t large_class)
+      {
+        real_state->push_large_stack(slab, large_class);
+      }
+
+      /**
+       * Reserve (and optionally commit) memory for a large sizeclass, proxies
+       * to the real implementation.
+       *
+       * This method must be implemented for `LargeAlloc` to work.
+       */
+      template<bool committed>
+      void* reserve(size_t large_class) noexcept
+      {
+        return real_state->template reserve<committed>(large_class);
+      }
+    };
+
+    /**
+     * Type for the allocator that exists inside the sandbox.
+     *
+     * Note that a real version of this would not have access to the shared
+     * pagemap and would not be used outside of the sandbox.
+     */
+    using InternalAlloc =
+      Allocator<never_init, no_op_init, MemoryProviderProxy>;
+
+    /**
+     * The start of the sandbox memory region.
+     */
+    void* start;
+
+    /**
+     * The end of the sandbox memory region
+     */
+    void* top;
+
+    /**
+     * State allocated in the sandbox that is shared between the inside and
+     * outside.
+     */
+    struct SharedState
+    {
+      /**
+       * The message queue for the allocator that lives outside of the
+       * sandbox but allocates memory inside.
+       */
+      struct RemoteAllocator queue;
+    } * shared_state;
+
+    /**
+     * The memory provider for this sandbox.
+     */
+    MemoryProviderStateMixin<NoOpPal> state;
+
+    /**
+     * The allocator for callers outside the sandbox to allocate memory inside.
+     */
+    ExternalAlloc alloc;
+
+    /**
+     * An allocator for callers inside the sandbox to allocate memory.
+     */
+    InternalAlloc* internal_alloc;
+
+    /**
+     * Constructor.  Takes the size of the sandbox as the argument.
+     */
+    Sandbox(size_t sb_size)
+    : start(alloc_sandbox_heap(sb_size)),
+      top(pointer_offset(start, sb_size)),
+      shared_state(new (start) SharedState()),
+      state(
+        pointer_offset(start, sizeof(SharedState)),
+        sb_size - sizeof(SharedState)),
+      alloc(state, SNMALLOC_DEFAULT_CHUNKMAP(), &shared_state->queue)
+    {
+      auto* state_proxy = static_cast<MemoryProviderProxy*>(
+        alloc.alloc(sizeof(MemoryProviderProxy)));
+      state_proxy->real_state = &state;
+      internal_alloc =
```

Comment:
I am not sure about directly allocating the allocator, rather than creating a pool of allocators, which this allocator is part of.  There are some operations that work over the set of allocators.  Mostly for debug, but directly allocating will mean that is not something we can do in the future.

Reply:
In the real implementation, we'll set up all of the alloc pool machinery inside the sandbox, this test is just checking that we can construct both kinds of allocator.  I've added a comment to this effect.

Note that in an environment with multiple trust domains we will never be able to enumerate *all* allocators.